sdks/go/pkg/beam/x/hooks/grpc/grpc.go

```diff
@@ -0,0 +1,76 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime/harness"
+	"github.com/apache/beam/sdks/go/pkg/beam/util/grpcx"
+	"google.golang.org/grpc"
+)
+
+var grpcHookKey = harness.HookKey("grpc")
+
+// Hook allow a runner to tailor various aspects of gRPC
+// communication with the FnAPI harness. Each member of the struct
+// is optional; the default behavior will be used if a value is not
+// supplied.
+type Hook struct {
```

Comment:
I was more expecting that most of these packages would be concrete hooks, rather than another layer of hook:

For example, a foo gRPC hook using a concrete "myDialer" Dialer could be a simple package:

func init() {
    harness.RegisterInitHook("foo", foo)
}

func foo(ctx context.Context) error {
       grpcx.Dialer = myDialer
       return nil
}

func Enable() {
      harnes.EnableHook("foo")
}

Or if we want opt-out instead simply:

func init() {
     harness.RegisterInitHook("foo", func(_ context.Context) error {
       grpcx.Dialer = myDialer
       return nil
    })
    harnes.EnableHook("foo")
}

If we need data, it can either be passes in the registration or -- if not available until later -- be added to the global options on Enable (and read back in the init hook at runtime).

Reply:
Your note about adding options on enable is reminiscent of my earlier note. I still prefer letting the user manage options in a delegated namespace rather than trying to make a generic argument to Enable.

The gRPC hook is not a good example here. There can only be one hook into that system. If two modules have different ideas on how to use gRPC, the system won't work.

However, the two-phase approach is useful in other circumstances, notably the perf package. In that case, we are mediating the usage of a global resource in the runtime. A runner or a DoFn may want to record a profile of an instruction execution. Those two applications can co-exist simultaneously. This hook structure allows for that to work. If we remove that, we remove this capability, which I believe is quite valuable. The discussion started from the premise that hooks help runners, but they can be helpful to other things as well, and I want to enable those use cases.

Reply:
Comments on the enablement API:

With the use of init() and calling harness.Enable(), this activates the hook based on importing the library. That doesn't seem like a good policy. Consider the perf hooks in the library. Having those hooks in the library is totally benign. I shouldn't have to alter my build to enable/disable performance benchmarking. The condition for enabling a hook is more fluid, and so should not be tied up in init, I feel.